lib/omnibus/file_syncer.rb

```diff
@@ -123,63 +124,102 @@ def sync(source, destination, options = {})
+      end
 
+      # Categorize files for processing
+      regular_files = []
+      symlinks = []
+      hardlinks = []
+
+      source_files.each do |source_file|
         case File.ftype(source_file).to_sym
         when :directory
-          create_sync_dir(File.join(source, relative_path), File.join(destination, relative_path))
+          # Skip - already created
         when :link
-          target = File.readlink(source_file)
-
-          Dir.chdir(destination) do
-            FileUtils.ln_sf(target, "#{destination}/#{relative_path}")
-          end
+          symlinks << source_file
         when :file
           source_stat = File.stat(source_file)
-          # Detect 'files' which are hard links and use ln instead of cp to
-          # duplicate them, provided their source is in place already
-          if hardlink? source_stat
-            if existing = hardlink_sources[[source_stat.dev, source_stat.ino]]
-              FileUtils.ln(existing, "#{destination}/#{relative_path}", force: true)
-            else
+          if hardlink?(source_stat)
+            hardlinks << [source_file, source_stat]
+          else
+            regular_files << source_file
+          end
+        else
+          raise RuntimeError,
+                "Unknown file type: `File.ftype(source_file)' at `#{source_file}'!"
+        end
+      end
+
+      # Process regular files and symlinks in parallel
+      parallel_items = regular_files + symlinks
+      thread_count = [8, parallel_items.size].min
+
+      if parallel_items.any?
+        ThreadPool.new(thread_count) do |pool|
+          regular_files.each do |source_file|
+            pool.schedule do
+              relative_path = relative_path_for(source_file, source)
               begin
                 FileUtils.cp(source_file, "#{destination}/#{relative_path}")
               rescue Errno::EACCES
                 FileUtils.cp_r(source_file, "#{destination}/#{relative_path}", remove_destination: true)
               end
-              hardlink_sources.store([source_stat.dev, source_stat.ino], "#{destination}/#{relative_path}")
             end
-          else
-            # First attempt a regular copy. If we don't have write
-            # permission on the File, open will probably fail with
-            # EACCES (making it hard to sync files with permission
-            # r--r--r--). Rescue this error and use cp_r's
-            # :remove_destination option.
-            begin
-              FileUtils.cp(source_file, "#{destination}/#{relative_path}")
-            rescue Errno::EACCES
-              FileUtils.cp_r(source_file, "#{destination}/#{relative_path}", remove_destination: true)
+          end
+
+          symlinks.each do |source_file|
+            pool.schedule do
+              relative_path = relative_path_for(source_file, source)
+              target = File.readlink(source_file)
+              FileUtils.ln_sf(target, "#{destination}/#{relative_path}")
```

Comment:
isn't it risky to create symlinks from within the same thread pool as regular files?